File: src/gpu/bitonicSort.ts
/**
 * PSCAD Modern - GPU Parallel Bitonic Sorting Engine
 * 
 * Executes O(log^2 N) parallel bitonic sorting for MMC submodule capacitor
 * voltage balancing across high-voltage converter arms (N = 100 .. 400+ SMs).
 */

export interface MmcSubmoduleState {
  smId: number;
  vCap: number;
  state: 'inserted' | 'bypassed' | 'blocked';
}

export class ParallelBitonicSorter {
  /**
   * Sort MMC submodule array using Bitonic Sorting Network
   * @param submodules Array of submodules with capacitor voltages
   * @param ascending If true, sort lowest to highest; if false, highest to lowest
   */
  static sortSubmodules(
    submodules: MmcSubmoduleState[],
    ascending: boolean = true
  ): MmcSubmoduleState[] {
    const n = submodules.length;
    if (n <= 1) return [...submodules];

    // Pad to next power of 2 for bitonic network
    const paddedLen = ParallelBitonicSorter.nextPowerOf2(n);
    const arr: MmcSubmoduleState[] = new Array(paddedLen);

    for (let i = 0; i < paddedLen; i++) {
      if (i < n) {
        arr[i] = { ...submodules[i] };
      } else {
        // Pad with sentinel
        arr[i] = {
          smId: 999999 + i,
          vCap: ascending ? Infinity : -Infinity,
          state: 'bypassed',
        };
      }
    }

    // Bitonic Sorting Stages
    for (let k = 2; k <= paddedLen; k <<= 1) {
      for (let j = k >> 1; j > 0; j >>= 1) {
        for (let i = 0; i < paddedLen; i++) {
          const l = i ^ j;
          if (l > i) {
            const blockAscending = ((i & k) === 0) ? ascending : !ascending;
            const shouldSwap = blockAscending
              ? arr[i].vCap > arr[l].vCap
              : arr[i].vCap < arr[l].vCap;

            if (shouldSwap) {
              const temp = arr[i];
              arr[i] = arr[l];
              arr[l] = temp;
            }
          }
        }
      }
    }

    // Strip padding
    return arr.filter(sm => sm.smId < 999999);
  }

  /**
   * Select submodules to insert/bypass based on MMC arm current and required levels
   * @param submodules Current submodule state array
   * @param numToInsert Number of submodules N_on to insert
   * @param armCurrent Instantaneous arm current i_arm
   */
  static balanceMmcArm(
    submodules: MmcSubmoduleState[],
    numToInsert: number,
    armCurrent: number
  ): MmcSubmoduleState[] {
    // If i_arm > 0 (charging), sort ascending to insert submodules with lowest Vc
    // If i_arm < 0 (discharging), sort descending to insert submodules with highest Vc
    const charging = armCurrent >= 0;
    const sorted = ParallelBitonicSorter.sortSubmodules(submodules, charging);

    const result: MmcSubmoduleState[] = [];
    for (let i = 0; i < sorted.length; i++) {
      const isInserted = i < numToInsert;
      result.push({
        smId: sorted[i].smId,
        vCap: sorted[i].vCap,
        state: isInserted ? 'inserted' : 'bypassed',
      });
    }

    // Restore original ordering by smId
    return result.sort((a, b) => a.smId - b.smId);
  }

  private static nextPowerOf2(n: number): number {
    let p = 1;
    while (p < n) {
      p <<= 1;
    }
    return p;
  }
}

File: src-tauri/src/gpu/bitonic_sort.rs
use serde::{Deserialize, Serialize};

/// MMC Submodule State
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SubmoduleVoltageState {
    pub sm_id: usize,
    pub v_cap: f64,
    pub is_inserted: bool,
}

/// Bitonic Sorter for Multi-Module MMC Capacitor Balancing
pub struct NativeBitonicSorter;

impl NativeBitonicSorter {
    /// In-place bitonic sort for MMC submodule voltages
    pub fn sort_submodules(submodules: &mut [SubmoduleVoltageState], ascending: bool) {
        let n = submodules.len();
        if n <= 1 {
            return;
        }

        let padded_n = n.next_power_of_two();
        let mut padded = Vec::with_capacity(padded_n);
        for sm in submodules.iter() {
            padded.push(sm.clone());
        }
        while padded.len() < padded_n {
            padded.push(SubmoduleVoltageState {
                sm_id: usize::MAX,
                v_cap: if ascending { f64::INFINITY } else { f64::NEG_INFINITY },
                is_inserted: false,
            });
        }

        let mut k = 2;
        while k <= padded_n {
            let mut j = k >> 1;
            while j > 0 {
                for i in 0..padded_n {
                    let l = i ^ j;
                    if l > i {
                        let block_asc = ((i & k) == 0) == ascending;
                        let should_swap = if block_asc {
                            padded[i].v_cap > padded[l].v_cap
                        } else {
                            padded[i].v_cap < padded[l].v_cap
                        };
                        if should_swap {
                            padded.swap(i, l);
                        }
                    }
                }
                j >>= 1;
            }
            k <<= 1;
        }

        // Copy back valid elements
        let mut valid_idx = 0;
        for sm in padded {
            if sm.sm_id != usize::MAX && valid_idx < submodules.len() {
                submodules[valid_idx] = sm;
                valid_idx += 1;
            }
        }
    }

    /// Balance MMC arm capacitor voltages
    pub fn balance_arm(submodules: &mut [SubmoduleVoltageState], n_on: usize, i_arm: f64) {
        let charging = i_arm >= 0.0;
        // If charging, sort ascending (lowest Vc first); if discharging, descending (highest Vc first)
        Self::sort_submodules(submodules, charging);

        for (idx, sm) in submodules.iter_mut().enumerate() {
            sm.is_inserted = idx < n_on;
        }

        // Restore original ID ordering
        submodules.sort_by_key(|sm| sm.sm_id);
    }
}
